### DISCUSSION TITLE
How would i extract types directly out of my introspection_types?

### DISCUSSION THREAD

**Discussion by stubifox** [0no-co/gql.tada#414] [Q&A] [answered]
Hi there,

I’m currently working with GraphQL Tada and I’m trying to directly extract types from the introspection_types. 

I understand that `ResultOf` and `VariablesOf` can be used for extracting types from queries and mutations, which I already use at some points but it does not help with my problem.

$\texttt{The following lines are just an example made up on the spot, I am trying to generalize the problem here :)}$

Lets say i have a query that retrieves `User` Data. The type of `User` is now located within the return Data somewhere.
We could say i could use  
```ts
type User = ResultOf<typeof USER_QUERY>['user']
```
Or something like that.
However my query has many possible return types encoded in `_typename`. Like `UserResponse` or `ErrorResponse` or `ServerInterceptionResponse`...
My TS Server can now not access `['user']`.

If I then want to declare a state/variable/Record with the type `User` I would need to copy the type from my backend or use codegen; which kind of defies the purpose of gql.tada.

I also tried to extract it from the introspection_types like this:
```ts
type User = introspection_types['User']['fields']
```
Which actually gives me the keys of the User but the types of the corresponding field of lets say `user.firstName` is still 
```
(property) 'firstName': {
    name: "firstName";
    type: {
        kind: "SCALAR";
        name: "String";
        ofType: null;
    };
}
```
Which i can not tranlate (easily) into typescript types. 

My question now is why it is not possible to easily extract types from the introspection_types and utilize them as typescript types. 

If there is a way please let me know. And thank you in advance.

Best regards,
Max

**Comment by kitten** [comment, marked as answer]
You can't retrieve a concrete type for an object type as it depends on a selection set, arguments, etc.

See:
- https://github.com/0no-co/gql.tada/discussions/233#discussioncomment-9165785
- https://discord.com/channels/1082378892523864074/1247901049294164018/1247904914957533215
- https://github.com/0no-co/gql.tada/discussions/381#discussioncomment-10463259

You should be creating fragments instead to derive the types of object selections

**Reply by stubifox** [reply]
thanks for the quick response!

**Reply by stubifox** [reply]
the discord link does not work for me though :/
